driver/src/main/java/org/neo4j/driver/internal/cursor/RxResultCursorImpl.java

```diff
@@ -47,428 +51,362 @@
-
-            if (recordConsumer != null) {
-                // records subscriber present
-                runnable = () -> {
-                    var closeStage = closeOnSummary ? boltConnection.close() : CompletableFuture.completedStage(null);
-                    closeStage.whenComplete((ignored, closeThrowable) -> {
-                        var error = Futures.completionExceptionCause(closeThrowable);
-                        if (error != null) {
-                            throwable.addSuppressed(error);
-                        }
-                        throwableConsumer.accept(throwable);
-                        recordConsumer.accept(null, throwable);
-                        summaryFuture.complete(summary);
-                        dispose();
-                    });
-                };
-            } else {
-                runnable = () -> {
-                    var closeStage = closeOnSummary ? boltConnection.close() : CompletableFuture.completedStage(null);
-                    closeStage.whenComplete((ignored, closeThrowable) -> {
-                        var error = Futures.completionExceptionCause(closeThrowable);
-                        if (error != null) {
-                            throwable.addSuppressed(error);
-                        }
-                        throwableConsumer.accept(throwable);
-                        summaryFuture.completeExceptionally(throwable);
-                        dispose();
-                    });
-                };
-            }
-        }
+    public synchronized Throwable getRunError() {
+        var name = runError == null ? "null" : runError.getClass().getCanonicalName();
+        log.trace("[%d] Run error explicitly retrieved (value=%s)", hashCode(), name);
+        runErrorExposed = true;
+        return runError;
+    }
 
-        runnable.run();
+    @Override
+    public List<String> keys() {
+        return runSummary.keys();
     }
 
     @Override
-    public void onIgnored() {
-        var throwable = termSupplier.get();
-        if (throwable == null) {
-            var message = "A message has been ignored during result streaming.";
-            throwable = new ClientException(
-                    GqlStatusError.UNKNOWN.getStatus(),
-                    GqlStatusError.UNKNOWN.getStatusDescription(message),
-                    "N/A",
-                    message,
-                    GqlStatusError.DIAGNOSTIC_RECORD,
-                    null);
-        }
-        onError(throwable);
+    public CompletionStage<Void> consumed() {
+        return consumedFuture;
     }
 
     @Override
-    public void onRecord(Value[] fields) {
-        var record = new InternalRecord(runSummary.keys(), fields);
-        synchronized (this) {
-            updateRecordState(RecordState.HAD_RECORD);
-            decrementDemand();
-        }
-        recordConsumer.accept(record, null);
+    public synchronized boolean isDone() {
```

Comment:
Minor thought. Perhaps this should be named isSuccessful or something like that as it only returns true on the SUCCEEDED case. Not important though.